goodtables/checks/duplicate_row.py

```diff
@@ -6,32 +6,38 @@
 
 import json
 from ..spec import spec
-from ..decorators import check
+from ..registry import check
 
 
 # Module API
 
-@check('duplicate-row')
-def duplicate_row(errors, columns, row_number, state):
-    rindex = state.setdefault('rindex', {})
-    try:
-        pointer = hash(json.dumps(list(column.get('value') for column in columns)))
-        references = rindex.setdefault(pointer, [])
-    except TypeError:
-        pointer = None
-    if pointer:
-        if references:
-            # Add error
-            message = spec['errors']['duplicate-row']['message']
-            message = message.format(
-                row_number=row_number,
-                row_numbers=', '.join(map(str, references)))
-            errors.append({
-                'code': 'duplicate-row',
-                'message': message,
-                'row-number': row_number,
-                'column-number': None,
-            })
-            # Clear columns
-            del columns[:]
-        references.append(row_number)
+@check('duplicate-row', type='structure', context='body')
+class DuplicateRow(object):
```

Comment:
Yes. I'll do it as a part of https://github.com/frictionlessdata/goodtables-py/issues/198. For now plan to finalize it based on `contrib.checks` implementation experience.